Sprint 5/step 1

Neverhood93 opened the pull request:
Доработана типизация.
Создан socketMiddleware, настроено socket-соединение с сервером.
Добавлен функционал на страницах Лента заказов и История заказов в профиле.
Добавлено модальное окно с детальной информацией о заказе.

Neverhood93 committed:
Добавлен универсальный socketMiddleware для подключения к WebSocket.
Добавлен feedSlice для получения всех заказов.
Изменена страница ленты заказов для тестирования WebSocket.

Neverhood93 committed:
Рефакторинг названий.

Neverhood93 committed:
Изменена маршрутизация для "/feed".
Добавлено подключение через WebSocket при монтировании компонента.

Neverhood93 committed:
Черновая верстка страницы Ленты заказов.

Neverhood93 committed:
Черновая верстка для маршрута "/profile/orders".
Также приведены в соответствие названия ранее реализованных страниц для единообразия.

Neverhood93 committed:
Создан profileOrdersSlice.
Добавлен profileOrdersMiddleware в store.

Neverhood93 committed:
Исправлен вызов WebSocket'ов.

Neverhood93 committed:
Добавлен балванка для модального окна с заказом, открытым из ленты.
Рефакторинг кода.

Neverhood93 committed:
Доработана логика в компоненте OrderDetails.

Neverhood93 committed:
Верстка детальной информации о заказе.

Neverhood93 committed:
Верстка карточки заказа в листе заказов.

Neverhood93 committed:
Черновая верстка компонента с номерами заказов в разных статусах.

Neverhood93 committed:
Верстка правой части страницы Ленты заказов.

Neverhood93 committed:
Добавлена проверка на существование ингредиента заказа в списке ингредиентов из стора.
Добавлено получение полной стоимости заказа.

Neverhood93 committed:
Рефакторинг кода.

Neverhood93 committed:
Доработано отображение ингредиентов конкретного заказа.

Neverhood93 committed:
Доработано отображение количества каждого ингредиента в детальной информации о заказе.

Neverhood93 committed:
Доработано отображение номеров заказов статусах: В работе и Готовы.

Neverhood93 committed:
Изменен socketMiddleware.

Neverhood93 committed:
Добавлена обработка случая получения ошибки "Invalid or missing token" в socket.

kyzinatra commented on src/services/middleware/socket-middleware.tsx:
```diff
@@ -0,0 +1,115 @@
+        clearTimeout(reconnectTimer);
+        isConnected = false;
+        reconnectTimer = 0;
+        socket.close();
```
Отлично! Отличная реализация сокетов.  Понравилось, что вы сделали reconnectTimer

kyzinatra commented on src/services/reducer.tsx:
```diff
@@ -4,11 +4,15 @@ import { selectedIngredientsSlice } from "./burger-constructor/reducer";
+import { feedSlice } from "./feed/reducer";
+import { profileOrdersSlice } from "./profile-orders/reducer";
 
 export const rootReducer = combineReducers({
```
Отлично! Хорошая типизация стора

kyzinatra commented on src/pages/feed-orders/feed-orders.tsx:
```diff
@@ -0,0 +1,48 @@
+
+function FeedOrdersPage() {
+  const FEED_SERVER_URL = "wss://norma.nomoreparties.space/orders/all";
+
```
Можно лучше: эту константу стоит вынести в отдельный файл, а сюда просто импортировать

kyzinatra commented on src/pages/feed-orders/feed-orders.tsx:
```diff
@@ -0,0 +1,48 @@
+  }, [dispatch]);
+
+  const filteredOrders = filterOrdersWithValidIngredients(orders, ingredients);
+
```
Можно лучше: тут можно использовать `useMemo()`

kyzinatra commented on src/components/order-list/order-list.tsx:
```diff
@@ -0,0 +1,21 @@
+        <ul className={styles.list}>
+          {data.map((item, index) => (
+            <OrderCard key={item._id} {...item} />
+          ))}
```
Надо исправить: прямым потомком `ul` может быть только `li`. В вашем случае это не так

kyzinatra commented on src/pages/profile-orders/profile-orders.tsx:
```diff
@@ -0,0 +1,57 @@
+    <main className={styles.page}>
+      <div className={styles.container_col}>
+        <ProfileNavBar />
+      </div>
```
Можно лучше: чтобы каждый раз не повторять `ProfileNavBar` стоит вынести компонент с шапкой и использовать `<Outlet />`

kyzinatra commented on src/pages/profile-orders/profile-orders.tsx:
```diff
@@ -0,0 +1,57 @@
+      dispatch(wsProfileOrdersDisconnect());
+    };
+  }, [dispatch]);
+
```
Можно лучше: стоит поправить ошибки линтера. Либо поправить правила конфига, либо сделать то, что просит линтер

kyzinatra commented on src/pages/profile-orders/profile-orders.tsx:
```diff
@@ -0,0 +1,57 @@
+function ProfileOrdersPage() {
+  const accessToken = localStorage.getItem("token") || "";
+  const PROFILE_ORDERS_SERVER_URL = `wss://norma.nomoreparties.space/orders?token=${accessToken.split(" ")[1]}`;
+
```
Можно лучше: Токен и url не стоит создавать в компоненте. Это не его задача

kyzinatra commented on src/pages/profile-orders/profile-orders.tsx:
```diff
@@ -0,0 +1,57 @@
+
+function ProfileOrdersPage() {
+  const accessToken = localStorage.getItem("token") || "";
+  const PROFILE_ORDERS_SERVER_URL = `wss://norma.nomoreparties.space/orders?token=${accessToken.split(" ")[1]}`;
```
Можно лучше: получение элемента из `localStorage` это эффект, который, по правилам реакта, нельзя использовать в теле компонента без `useEffect`

kyzinatra commented:
Выглядит очень круто! Видно, что проделана большая работа. Но все же есть моменты на которые стоит обратить внимание.
Подробнее: 

**Что важно исправить:**

-  прямым потомком `ul` может быть только `li`. В вашем случае это не так


**Можно лучше:**

-  эту константу стоит вынести в отдельный файл, а сюда просто импортировать
-  тут можно использовать `useMemo()`
-  чтобы каждый раз не повторять `ProfileNavBar` стоит вынести компонент с шапкой и использовать `<Outlet />`
-  стоит поправить ошибки линтера. Либо поправить правила конфига, либо сделать то, что просит линтер
-  Токен и url не стоит создавать в компоненте. Это не его задача
-  получение элемента из `localStorage` это эффект, который, по правилам реакта, нельзя использовать в теле компонента без `useEffect`


**Что понравилось:**

-  Отличная реализация сокетов.  Понравилось, что вы сделали reconnectTimer
-  Хорошая типизация стора


**Обратите внимание, что работа принимается только после исправления всех `«Надо исправить»`.**

Neverhood93 committed:
Исправление замечаний:
Вынесение констант в отдельный файл.
Добавлен useMemo в функции фильтрации заказов.
Изменен компонент OrderList, чтобы элементы ul были li.

Neverhood93 committed:
Исправление замечаний:
Получение элемента из localStorage это эффект, который, по правилам реакта, нельзя использовать в теле компонента без useEffect.
